### PR TITLE
draw: fix dotted line rendering when zoomed out

### PR DESCRIPTION
**Pull request by mimecuvalo**
Fixes https://github.com/tldraw/tldraw/issues/1995

### Change type

- [x] `bugfix`
- [ ] `improvement`
- [ ] `feature`
- [ ] `api`
- [ ] `other`

### Release notes

- Draw: fix dotted line shape rendering when zoomed out greatly.

### CODE REVIEW

File: packages/tldraw/src/lib/shapes/draw/getPath.ts [hunk 1]
```diff
@@ -119,11 +119,17 @@ export function getPointsFromSegments(segments: TLDrawShapeSegment[]) {
...
+	strokeWidth: number,
+	zoomLevel: number
+) {
 	return {
 		draw: 'none',
 		solid: `none`,
-		dotted: `0.1 ${strokeWidth * 2}`,
+		// If we're zoomed way out (10%), then we need to make the dotted line go to 9 instead 0.1
+		// Chrome doesn't render anything otherwise.
+		dotted: `${zoomLevel < 0.2 ? 0 : 0.1} ${strokeWidth * 2}`,
```
Comment (MitjaBezensek) [inline packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: Seems like it also disappears at 21 and sometimes even 22%? At least from my tests 23% works.

https://github.com/user-attachments/assets/6ad40f86-a9fc-404a-aebc-69c41ead9666






```suggestion
		dotted: `${zoomLevel < 0.23 ? 0 : 0.1} ${strokeWidth * 2}`,
```

**Review by MitjaBezensek — APPROVED** [review]
Nice! 👏 

Found a small issue with some percentages around 20%, left a comment.

File: packages/tldraw/src/lib/shapes/draw/getPath.ts (hunk 1 above)
Comment (steveruizok) [reply packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: Browser scale is going to effect this too :(

Comment (mimecuvalo) [reply packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: > Seems like it also disappears at 21 and sometimes even 22%? At least from my tests 23% works.

argh, that's worrisome then, that isn't consistent with when it's appearing/disappearing for me. fun fun fun.
@MitjaBezensek is that when it's on the non-retina screen maybe?

Comment (MitjaBezensek) [reply packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: Ah yes. On the laptop screen it seems at around 10% and on external, non-retina one its at 22% 🤷‍♂️

**Comment by steveruizok** [comment]
If we can just improve it at 100% zoom then let's land this. Zoomers are beyond the pale

**Comment by steveruizok** [comment]
Added fixes across all other shapes when using perfect dash props. Also deleted duplicated code for perfect dash props!

File: packages/tldraw/src/lib/shapes/draw/getPath.ts (hunk 1 above)
Comment (steveruizok) [reply packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: Settling on 20%. we can dial this in

Comment (mimecuvalo) [reply packages/tldraw/src/lib/shapes/draw/getPath.ts:132]: nice! but yeah, if it's still happening from 20 < x < 23, how can else we improve this then?

**Pull request merged**